### ISSUE TITLE
Bug in tf.einsum - Returns different values from np.einsum for identical parameters

### ISSUE THREAD

**Issue by honzukka** [tensorflow/tensorflow#29869]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or binary): Binary (Anaconda)
- TensorFlow version (use command below): 1.13.1
- Python version: 3.7.3
- CUDA/cuDNN version: 10.0/7.6.0
- GPU model and memory: NVIDIA Titan X (Pascal), 12 GB

**Describe the current behavior**

`tf.einsum` returns buggy values compared to `np.einsum` for identical parameters. Here is the current output (please find code below):

```
np: 8.610251426696777
tf: 0.0
```

Please also note that the same code works as expected when running on the CPU. It also works correctly on the GPU in TF 1.12.0, CUDA/cuDNN 9.0/7.3.1 and Python 3.6.8.

**Describe the expected behavior**

Here is the expected output:

```
np: 8.610251426696777
tf: 8.610251426696777
```

**Code to reproduce the issue**

```
from __future__ import (
    absolute_import,
    division,
    print_function
)

import tensorflow as tf
import numpy as np

tf.enable_eager_execution()

einsum_string = "bijk,bijk->bij"

# load values
S_h_numpy = np.load("./S_h.npy")
S_h = tf.convert_to_tensor(S_h_numpy, dtype=tf.float32)

h_numpy = np.load("./h.npy")
h = tf.convert_to_tensor(
    h_numpy,
    dtype=tf.float32
)

# perform einsum
ht_S_h_numpy = np.einsum(einsum_string, h_numpy, S_h_numpy)
ht_S_h = tf.einsum(einsum_string, h, S_h)

print("np: {np_val}\ntf: {tf_val}".format(
    np_val=ht_S_h_numpy[2, 191, 191],
    tf_val=ht_S_h[2, 191, 191],
))
```

**Other info / logs**

Here are the input tensors used in the example: [input_tensors.zip](https://github.com/tensorflow/tensorflow/files/3296468/input_tensors.zip).

The tensors are of shape (4, 192, 192, 2). I could not choose a smaller example because the buggy values appear only in part of the second and all of the following batch items. The bug therefore seems to be connected to the size of the input.

**Comment by honzukka** [comment]
I have to add that after updating my GPU drivers and disabling the eager mode, the code seems to execute fine. I don't use the eager mode by default, only for debugging. Does this mean that I shouldn't use it at all?

**Comment by gadagashwini** [comment]
I am able to reproduce the issue on colab with Tensorflow 1.13.1. Thanks!

**Comment by alextp** [comment]
I can confirm I can reproduce a difference between tf and numpy only on GPU for tensors of that shape, even without using the specific values. Running on the CPU makes the problem go away.

My repro (on tf-nightly-gpu-2.0-preview) is

```
import tensorflow as tf
import numpy as np


einsum_string = "bijk,bijk->bij"

# load values
S_h_numpy = np.random.random([4, 192, 192, 2]).astype(np.float32)
S_h = tf.convert_to_tensor(S_h_numpy, dtype=tf.float32)

h_numpy = np.random.random([4, 192, 192, 2]).astype(np.float32)
h = tf.convert_to_tensor(
    h_numpy,
    dtype=tf.float32
)

# perform einsum
with tf.device('cpu:0'):
  ht_S_h_numpy = np.einsum(einsum_string, h_numpy, S_h_numpy)
  ht_S_h = tf.einsum(einsum_string, h, S_h)

print("np: {np_val}\ntf: {tf_val}".format(
    np_val=ht_S_h_numpy[2, 191, 191],
    tf_val=ht_S_h[2, 191, 191],
))
```

I think then this must be a bug in the GPU kernels.

@rmlarsen do you know who is the best person to look at this?

**Comment by hstuk** [comment]
Is there any progress on that bug? I think I have a similar problem, here:

https://github.com/tensorflow/tensorflow/issues/31022

**Comment by alextp** [comment]
Escalating to @tatianashp

**Comment by rmlarsen** [comment]
Anudhyan, would you have time to take a look at this?

**Comment by bloops** [comment]
Sorry, I don't have spare cycles currently.

However, this is still using the old tf.einsum and not the new EinsumOp. Which means that the graph would have exactly one tf.matmul (BatchMatMul) and one or two Transpose ops. So either of those two GPU kernels would be the source of the problem.

**Comment by rmlarsen** [comment]
Sorry, I am really running short on cycles this month :(

However, this is still using the old tf.einsum and not the new EinsumOp.
The problem is then likely in the GPU kernel for BatchMatMul (which is a
bit surprising, as it's been around for while).

**Comment by rmlarsen** [comment]
No worries. I'll take a look.

**Comment by bloops** [comment]
I haven't tried this, but I think in theory, the problem should also be reproducible with:
  a = np.random.random([147456, 1, 2]).astype(np.float32)   # 147456 = 4 * 192 * 192
  b = np.random.random([147456, 2, 1]).astype(np.float32)
  c = tf.matmul(a, b)

**Comment by hstuk** [comment]
i checked this code and the difference is zero.

`
import tensorflow as tf
import numpy as np

print(tf.__version__)

@tf.function
def mul(a,b):
    return tf.matmul(a,b)

a = np.random.random([147456, 1, 2]).astype(np.float32)  # 147456 = 4 * 192 * 192
b = np.random.random([147456, 2, 1]).astype(np.float32)

conf = tf.ConfigProto()
conf.gpu_options.allow_growth = True
sess = tf.Session(config=conf)

with tf.device("/cpu:0"):
    c = mul(a,b).eval(session=sess)

with tf.device("/gpu:0"):
    c_gpu = mul(a,b).eval(session=sess)

print('difference', np.sum(np.abs(c-c_gpu)))
`

**Comment by hstuk** [comment]
are there any news on this bug?

**Comment by bloops** [comment]
I think this bug is a duplicate of this GPU Matmul kernel issue:
https://github.com/tensorflow/tensorflow/issues/26969

I could reproduce this with matmul as well. Both the previous Python implementation of tf.einsum and the new implementation with EinsumOp  (submitted 09/27) uses the same GPU kernels as matmul.

Below is my reproduction on `tf-nightly-gpu-2.0-preview`.

```
with tf.compat.forward_compatibility_horizon(2019, 10, 20):
  n = 2**16  # gives correct answer with n = 2**16 - 1.
  a = np.random.random([n, 1, 1]).astype(np.float32)
  b = np.random.random([n, 1, 1]).astype(np.float32)

  with tf.device('gpu:0'):
    x = np.matmul(a, b)
    y = tf.matmul(a, b)
    z = tf.einsum("Aij,Ajk->Aik", a, b)

  print("np: {}\ntf1: {}\ntf2: {}".format(
      x[n - 1,0,0],
      y[n - 1,0,0],
      z[n - 1,0,0]
  ))
```

**Comment by hoangcuong2011** [comment]
Is there any update on this bug? I personally think einsum is a very powerful operation, and people might use it a lot. It would be not safe at all if we use it while knowing there exists a certain bug that makes it failed to work properly.

**Comment by bloops** [comment]
Looks like referred GPU issue was closed as not reproducible in 2.1.0-dev20191226. Is the einsum problem still happening?

**Comment by tensorflowbutler** [comment]
Hi There,
 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 
 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
